app/oidc/oidc.py

```diff
@@ -0,0 +1,66 @@
+import functools
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta, timezone
+from typing import Callable, ParamSpec
+
+import pytz
+from google.auth.credentials import Credentials
+from google.auth.transport.requests import Request
+from structlog import get_logger
+
+from app.settings import OIDC_TOKEN_LEEWAY_IN_SECONDS
+
+P = ParamSpec("P")
+
+logger = get_logger()
+TTL = 3600 - OIDC_TOKEN_LEEWAY_IN_SECONDS  # 1 hour minus leeway in seconds
+
+
+def get_expiry_from_ttl(ttl: int) -> datetime:
+    return datetime.now(timezone.utc) - timedelta(seconds=ttl)
+
+
+def is_naive(dt: datetime) -> bool:
+    return dt.tzinfo is None or dt.tzinfo.utcoffset(dt) is None
+
+
+def refresh_oidc_credentials(
+    func: Callable[P, Credentials]
+) -> Callable[P, Credentials]:
+    """
+    This decorator is used on implementations of the abstract method OIDCCredentialsService.get_credentials.
+    This decorated method uses a ttl_cache decorator to discard fetched credentials after a given time.
+    However, the TTL of the credentials should be determined by inspecting the expiry of the credentials
+     themselves, not relying on the ttl_cache.
+    As such, this decorator ensures the credentials returned from the ttl_cache are valid and refreshed
+     via inspecting the expiry property of the credentials.
+    Overall this ensures the caller is always returned valid credentials and credentials are cached to minimise
+     API calls for credentials as well as discarded once invalid.
+    """
+
+    @functools.wraps(func)
+    def wrapper_refresh_oidc_credentials(
+        *args: P.args, **kwargs: P.kwargs
+    ) -> Credentials:
+        credentials = func(*args, **kwargs)
+        expiry = (
+            pytz.utc.localize(dt=credentials.expiry)
+            if is_naive(credentials.expiry)
+            else credentials.expiry
+        )
```

Comment:
We don't need to localise it but I'm not sure what it is conditional anyway, the lib should be consistent as it is either timezone aware or not.

But either way, you can just add the timezone to the datetime instead of localising it as the lib always passes a UTC date regardless of it is it naive, therefore you can just call `.replace(tzinfo=timezone.utc)` on the date.

```py
expiry = credentials.expiry.replace(tzinfo=timezone.utc)
if expiry < get_expiry_from_ttl(TTL):
   ...
```

Reply:
Annoyingly its the google auth lib which gives a utc datetime for the expiry on credentials, but its naive. So to do the calculations I needed to add it, we can replace like you say, or localize it.
The reason I opted to localize is to guarantee the decorator can handle the credentials as it will error if it tries to calculate validity with naive and aware dates.

Reply:
I understood why you needed to localise but not why it was conditional `if naive` (it should always be naive or not). Was it because the local one wasn't naive?.

The reason I suggested `.replace` is, localise implies it could be timezone aware which isn't utc but we want to convert to utc, but that is never the case. So since it is always utc, we just need to add tzinfo, for the sake of comparing two timezone aware objects.

Reply:
Oh wait sorry you weren't meaning to move it out of the decorator - ignore me, will use replace each time instead of check naivety and localize 👍

Reply:
Done 👍